### PR TITLE
18MEX: Support NdM merge

### PR DESCRIPTION
**Pull request by perwestling**
When first 5 train is bought the NdM merge event is triggered
if NdM has floated.

The merge process if that player in order, left of the player
that has bought the 5 train, is asked if their corporation wants
to merge. (Excluding "American" corporations, and corporations
controlled by NdM president.) If there is no floated corporation
that accepts, NdM will merge with any non floated, if there is
any eligible one.
The president of the merged corporation receives a 10% reserved
NdM share in exchange for their presidency. Remaining shares
in merged corporation is force sold for half amrket price.
Treasury and trains are transfered to NdM.
NdM also exchanges up to two tokens, one should at least be
the home token if any.

There are a new step: merge, which is located in module 18MEX.
This is used for the merge question. A connected view has
been added for the merge step.

18MEX also uses an assign step which is used in case NdM need
to choose between several tokens on the map.

---------

Have added a buyable attribute to share. This is set to true by
default, and is set to false if share reserved. In 18MEX this
is done for exchange shares that will be used during minor
closure and NdM merge.

### CODE REVIEW

File: assets/app/view/game/merge.rb
Comment (jenf) [inline assets/app/view/game/merge.rb:30]: Bit confused on this, should probably be a needs into the view to say the player name.

File: assets/app/view/game/merge.rb
Comment (jenf) [inline assets/app/view/game/merge.rb:22]: Should be generic and have NdM as a const.

File: assets/app/view/game/merge.rb
Comment (jenf) [inline assets/app/view/game/merge.rb:8]: Is there any way we can get 1817 to use some of this code?

**Comment by perwestling** [comment]
Here is how the Merge view looks like when a player is asked if their corporation should be merged into NdM:
image

And this is how it looks when a non-floated corporation should be merged into NdM:
image

And here is how it looks like when CHI accepts the merge, and NdM has to select which token (besdies CHI home) that should be swapped (the non-chosen is removed):
image

**Comment by perwestling** [comment]
That dialog that let's NdM confirm that SPM should be merged could instead be automatic - if the last candidate is a non-floated one it should be selected. The only case for NdM to have to do a choice should be if there are several non-floated candidates that have player owned 20% shares; which is very unlikely to happen.

BTW, this is how it looks like if NdM player clicks on Mazatlán (removing CHI token in Torreón):
image

Note that CHI has disappeared from the order. And in the Entities view all CHI has been removed:
image

File: assets/app/view/game/merge.rb
Comment (perwestling) [reply assets/app/view/game/merge.rb:30]: See pictures below. The idea here is the merge candidate is shown, and the president of it makes the decision if it should merge. In case it has no president, it is a non-started one, and then NdM should make the decision. (Note my remark below that this is very likely automatic.)

I know realize that this is probably an error - only if the corporation is floated the player can decide. Non-floated is automatically merged if one if players decline. If non-floated and non-sold presidency, that is chosen if players decline. And if there is two non-floated eligible (both non-sold, or sold) - then NdM gets to choose.

Comment (perwestling) [reply assets/app/view/game/merge.rb:30]: As can be seen from the screen shots, the merge candidate is shown underlined in the Order, as well as in text above the button. Only time it is not shown underlined is if it is a non-floated corporation that is up for merge.

File: assets/app/view/game/merge.rb
Comment (perwestling) [reply assets/app/view/game/merge.rb:22]: That's an idea. I need to make @game.ndm generic as well. Probably add a method to base that g_18_mex.rb can override?  What do you call the one that you merge into? Mergee for the one that is merged into the... Merger?

Or should this view be only used by 18MEX? Can it be put in a Module in that case?

File: assets/app/view/game/merge.rb
Comment (perwestling) [reply assets/app/view/game/merge.rb:8]: I have not studied 1817 in detail but it would be good if it was. I don't recall such a situation in 1817, where there is a list of candidate corporations to be merged into a target, but maybe there is? 
I suppose it could be used for a Merge/Decline question of one mergee into one merger.

Comment (jenf) [reply assets/app/view/game/merge.rb:8]: In acquisitions once a player has won the bid they then choose the corporation to merge with it.

Comment (perwestling) [reply assets/app/view/game/merge.rb:8]: Is there such a dialog implemented today? If not it sounds like this can be generalized to cover for that as well.  
Will try to generalize it a bit.

File: assets/app/view/game/corporation.rb
Comment (tobymao) [inline assets/app/view/game/corporation.rb:351]: index i believ is the same

File: assets/app/view/game/merge.rb
Comment (tobymao) [reply assets/app/view/game/merge.rb:22]: the views shouldn't have any game specific logic like this

File: lib/engine/game/g_18_mex.rb
Comment (tobymao) [inline lib/engine/game/g_18_mex.rb:183]: unless major

Comment (perwestling) [reply lib/engine/game/g_18_mex.rb:183]: Done

File: assets/app/view/game/merge.rb
Comment (perwestling) [reply assets/app/view/game/merge.rb:22]: Ok, I have now generalized the view so it can be used by other titles.

File: assets/app/view/game/merge.rb
Comment (perwestling) [reply assets/app/view/game/merge.rb:30]: Improved the code a bit to make it clearer.

File: assets/app/view/game/corporation.rb
Comment (perwestling) [reply assets/app/view/game/corporation.rb:351]: Yes, index is just an alias to find_index. I changed to use index in both lines now as it is shorter.

**Comment by perwestling** [comment]
@jenf and @tobymao : I have made some improvements, and handled review comments.

One thing I added was to display the reserved (in 18MEX: trade-in) IPO shares in the corporation charter. These cannot be bought. (The title should ensure these are the last ones in the share array so that all other shares can be bought.)
Here is a screen shot on how this looks from the start in 18MEX:
image

And here how it looks just before the NdM merge (when the last trade-in share is "up for grabs")
image

Regardless if merge take place or not that last share is changed to buyable, so the "Trade-in shares" section disappear from NdM. As can be seen CHI does not have any trade-in shares at all. In 18MEX NdM has 3 and UdY has 1 from the start.

File: assets/app/view/game/merge.rb
Comment (perwestling) [reply assets/app/view/game/merge.rb:8]: Have generalized it a bit. But still not sure if it is possible to use it in 1817. Maybe in 1824?

File: assets/app/view/game/merge.rb
Comment (jenf) [inline assets/app/view/game/merge.rb:21]: This is a bit of a different flow than normal, but we are hitting issues where the 'active' player is not one that's normally listed on the round including in 1817.

Maybe we should have a standard way of showing that the normal turn sequence flow has been interrupted and who the interrupting entity is. But still showing the normal turn sequence as before.

@tobymao this would be nice to include on 1817 as the acquisiton is player based, but when your gaining/paying off loans it's corporation based. Would also help on the cash crisis as well, as it will get confusing whose turn it is.

Comment (jenf) [reply assets/app/view/game/merge.rb:21]: The current entity should be the 'merge_decider' anyway. So that the email gets sent to the right user.

Comment (perwestling) [reply assets/app/view/game/merge.rb:21]: In g_18_mex/merge.rb, active_entities is [@game.merge_decider].
Normally, the merge_decider in 18MEX would be the president of the corporation to be merged into NdM. But there is one exception - if there is two(+) eligible corporations, that has not floated, then NdM will be the merge decider. It is probably a rare case.
If there is just one non-floated to choose from NdM is not asked but instead it is auto merged.

But NdM might have to select among multiple tokens placed, and that will also interrupt normal flow, before returning to Buy Trains step.

Comment (jenf) [reply assets/app/view/game/merge.rb:21]: Merging up to the latest should add the current entity to the list, removing the need for the "To make the decision" part

**Comment by perwestling** [comment]
Have removed the "To make the decision" part, moved the "Merge candidates" to the right of the Merge button. Rebased versus master.

Ready to be reviewed again.

File: assets/app/view/game/merge.rb
Comment (tobymao) [reply assets/app/view/game/merge.rb:8]: has this been attempted yet? that's the only thing i'd look into for this commit

Comment (perwestling) [reply assets/app/view/game/merge.rb:8]: I have not done any 1817 related work. So maybe @tobymao or @jenf could try to use it for acquisitions?

File: assets/app/view/game/merge.rb
Comment (perwestling) [reply assets/app/view/game/merge.rb]: Have removed the "To make the decision" part, moved the "Merge candidates" to the right of the Merge button. Rebased versus master.

Have not uploaded new screen shots as the changes are very minor.

**Comment by perwestling** [comment]
I will resolve the conflicts, but I am waiting with that until it has been enough reviewed.

**Comment by perwestling** [comment]
@tobymao and @jenf I have now updated this commit. Have rebased it versus latest master.
In the commit it now uses the already existing merge dialog. That had to be updated a bit to make it work for 18MEX.
I also introduced preselected corporation in case only one candidate exists (which is the case for 18MEX where one candidate at a time is presented as a choice).

**Comment by perwestling** [comment]
I have now fixed the exchange tokens (so that they appear on the charter only as part of the NdM merge.

So with this PR I think 18MEX is ready, functionally, to be put to alpha. I am doing some pre-alpha testing on the test site. It uses the same code, except the possibility to inter-buy corporations (as it breaks the started game if added). 

So it would be good to have it reviewed.

File: assets/app/view/game_page.rb
Comment (tobymao) [inline assets/app/view/game_page.rb:254]: current_actions.include?('merge')

**Review by tobymao — APPROVED** [review]

File: assets/app/view/game_page.rb
Comment (perwestling) [reply assets/app/view/game_page.rb]: Fixed it

**Comment by perwestling** [comment]
@tobymao I haved fixed your comment. Time to merge it?

**Pull request merged**
